cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+"""
+Copyright 2023 Impulse Innovations Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy
+
+from cai_causal_graph import CausalGraph
+from cai_causal_graph.graph_components import Edge, Node, get_name_from_lag, get_variable_name_and_lag
+from cai_causal_graph.interfaces import HasIdentifier
+from cai_causal_graph.type_definitions import EDGE_T, NodeLike, NodeVariableType
+
+logger = logging.getLogger(__name__)
+
+# TODO: we can do one general for many other things as well
+def reset_attributes(func):
+    """Decorator to reset the summary graph."""
+
+    # TODO: make this more clever as it is not said that we need to reset the summary graph etc
+    def wrapper(self, *args, **kwargs):
+        function = func(self, *args, **kwargs)
+        self._summary_graph = None
+        self._order = None
+        self._variables = None
+        return function
+
+    return wrapper
+
+
+def extract_names_and_lags(
+    node_names: List[NodeLike],
+) -> Tuple[List[Dict[str, int]], int]:
+    """
+    Extract the names and lags from a list of node names.
+    This is useful for converting a list of node names into a list of node names and lags.
+
+    Example:
+        ['X', 'Y', 'Z lag(n=2)'] -> [{'X': 0}, {'Y': 0}, {'Z': 2}]
+
+    :param node_names: List of node names.
+    :return: List of dictionaries with node names and lags sorted by lag and the maximum lag.
+    """
+    names_and_lags = []
+    for node_name in node_names:
+        variable_name, lag = get_variable_name_and_lag(node_name)
+        names_and_lags.append({variable_name: lag})
+    sorted_names_and_lags = sorted(names_and_lags, key=lambda x: max(x.values()), reverse=False)
+    maxlag = next(iter(sorted_names_and_lags[0].values()))
+    return sorted_names_and_lags, maxlag
+
+
+class TimeSeriesCausalGraph(CausalGraph):
+    """
+    A causal graph for time series data.
+
+    The node in a time series causal graph will have additional metadata that
+    gives the time information of the node together with the variable name.
+    The two additional metadata are:
+    - 'timedelta': the time difference with respect the reference time 0
+    - 'variable_name': the name of the variable (without the lag information).
```

Comment:
I would urge you to provide an example usage snippet below. We should really be aiming to have good doc strings showing examples etc.